neqo-http3/src/connection.rs

```diff
@@ -1114,38 +1090,41 @@ impl Http3Events {
         self.insert(Http3Event::StateChange(state));
     }
 
-    pub fn events(&self) -> BTreeSet<Http3Event> {
-        self.replace(BTreeSet::new())
-    }
-
-    pub fn replace(&self, new_events: BTreeSet<Http3Event>) -> BTreeSet<Http3Event> {
-        self.events.replace(new_events)
+    pub fn events(&self) -> impl Iterator<Item = Http3Event> {
+        self.events.replace(BTreeSet::new()).into_iter()
     }
 
     fn insert(&self, event: Http3Event) {
         self.events.borrow_mut().insert(event);
     }
 
+    fn remove(&self, event: &Http3Event) -> bool {
+        self.events.borrow_mut().remove(event)
+    }
+
     fn remove_events_for_stream_id(&self, remove_stream_id: u64) {
-        let updated_events = self
+        let events_to_remove = self
             .events
-            .replace(BTreeSet::new())
+            .borrow()
             .iter()
             .filter(|evt| match evt {
                 Http3Event::HeaderReady { stream_id }
                 | Http3Event::DataWritable { stream_id }
                 | Http3Event::DataReadable { stream_id }
                 | Http3Event::NewPushStream { stream_id }
                 | Http3Event::Reset { stream_id, .. }
-                | Http3Event::StopSending { stream_id, .. } => *stream_id != remove_stream_id,
+                | Http3Event::StopSending { stream_id, .. } => *stream_id == remove_stream_id,
                 Http3Event::AuthenticationNeeded
                 | Http3Event::GoawayReceived
                 | Http3Event::StateChange { .. }
-                | Http3Event::RequestsCreatable => true,
+                | Http3Event::RequestsCreatable => false,
             })
             .cloned()
-            .collect::<BTreeSet<_>>();
-        self.events.replace(updated_events);
+            .collect::<SmallVec<[_; 8]>>();
+
+        for evt in events_to_remove {
```

Comment:
I still think that [HashSet.retain](https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.retain) is a better choice here.

Reply:
we need an answer for the ordering issue to switch to HashSet.